libs/sdk-core/src/persist/swap_taproot.rs

```diff
@@ -0,0 +1,433 @@
+                AND sp.tx_id = r.spent_tx_id 
+                AND sp.output_index = r.spent_output_index
+            LEFT JOIN payments p ON s.payment_hash = p.id
+            LEFT JOIN sync.open_channel_payment_info o ON s.payment_hash = o.payment_hash
+            LEFT JOIN taproot_swap_cache c ON s.address = c.address
+            ") + where_clause + 
+            " ORDER BY created_at DESC"),
+        )?;
+        let mut rows = query.query(params)?;
+        let swaps = self.map_full_taproot_swap_rows(&mut rows)?;
+        Ok(swaps)
+    }
+
+    fn map_full_taproot_swap_rows(
+        &self,
+        rows: &mut Rows<'_>,
+    ) -> Result<Vec<FullTaprootSwapData>, PersistError> {
+        let mut result = HashMap::new();
+        while let Some(row) = rows.next()? {
+            let address: String = row.get("address")?;
+            if !result.contains_key(&address) {
+                let payment_hash: String = row.get("payment_hash")?;
+                let payment_hash = hex::decode(payment_hash)?;
+                let accepted_opening_fee_params: String = row.get("accepted_opening_fee_params")?;
+                let accepted_opening_fee_params =
+                    serde_json::from_str(&accepted_opening_fee_params)?;
+                let max_swap_amount_sat: Option<u64> = row.get("max_swap_amount_sat")?;
+                let parameters = match max_swap_amount_sat {
+                    Some(max_swap_amount_sat) => Some(TaprootSwapParameters {
+                        max_swap_amount_sat,
+                        min_swap_amount_sat: row.get("min_swap_amount_sat")?,
+                        min_utxo_amount_sat: row.get("min_utxo_amount_sat")?,
+                    }),
+                    None => None,
+                };
+                result.insert(
+                    address.clone(),
+                    FullTaprootSwapData {
+                        swap: TaprootSwap {
+                            address: address.clone(),
+                            created_at: row.get("created_at")?,
+                            lock_time: row.get("lock_time")?,
+                            preimage: row.get("preimage")?,
+                            payment_hash,
+                            refund_private_key: row.get("refund_private_key")?,
+                            claim_public_key: row.get("claim_public_key")?,
+                            accepted_opening_fee_params,
+                        },
+                        outputs: Vec::new(),
+                        paid_amount_msat: row.get("paid_amount_msat")?,
+                        bolt11: row.get("bolt11")?,
+                        last_payment_error: row.get("last_payment_error")?,
+                        parameters,
+                        refund_transactions: Vec::new(),
+                    },
+                );
+            }
+
+            let current_swap = result
+                .get_mut(&address)
+                .expect("just inserted key is not in hashmap");
```

Comment:
Can we handle the error instead of panicking?

Reply:
I agree you shouldn't panic. But in this case the item is definitely in the hashmap?